### PR TITLE
Add PipelineRuns list API

### PR DESCRIPTION
**Pull request by JohnNiang**
### What this PR dose

- Remove unused GVRs that have registered into informer
- Refine GroupVersionResource construction
- Remove unused dependency: github.com/prometheus/client_golang@v1.8.0
- Remove useless replacement in go.mod
- Add client and replace cache for APIServer
- Add PipelineRuns list API
- Add backward parameter for PipelineRuns list API

### Why we need it

Before this, if we want to use clients of our CRDs, we have to execute command `make clientset` to generate some boilerplate codes.

For this PR, I include [controller-runtime client](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client.go) to let us use clients conveniently. Below is an example:

```go
var prs v1alpha4.PipelineRunList
if err := o.Client.List(context.Background(), &prs); err != nil {
	return
}
```

This PR is to prepare for `PipelineRun feature in APIServer`. At the same time, this PR is also a part of #41.

### Special notes

Docker image for test: `johnniang/devops-apiserver:add-client-and-replace-cache`

/kind feature

### CODE REVIEW

File: pkg/kapis/devops/v1alpha4/register.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	K8sClient k8s.Client
+	Client    client.Client
+}
+
+func AddToContainer(o Option) error {
+	// TODO Complete definitions of route
+	//ws := runtime.NewWebService(v1alpha4.GroupVersion)
+	//ws.Route(ws.GET("/devops/{devops}/credentials"))
+	ws := runtime.NewWebServiceWithoutGroup(v1alpha4.GroupVersion)
+	// TODO Remove this test API
```
Comment (LinuxSuRen) [inline pkg/kapis/devops/v1alpha4/register.go:25]: Are you sure this PR is ready?

Comment (JohnNiang) [reply pkg/kapis/devops/v1alpha4/register.go]: hi @LinuxSuRen , this PR has ready for review. Thank you for your patience.

File: pkg/kapis/devops/v1alpha4/handler.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+			transformFunc = backwardTransform()
+		}
+
+		apiResult := v1alpha32.DefaultList(convertPipelineRunsToObject(prs.Items), queryParam, v1alpha32.DefaultCompare(), v1alpha32.DefaultFilter(), transformFunc)
+		_ = response.WriteAsJson(apiResult)
+	}
+}
+
+// backwardTransform transforms PipelineRun into JSON raw message of Jenkins run status.
+func backwardTransform() v1alpha32.TransformFunc {
```
Comment (LinuxSuRen) [inline pkg/kapis/devops/v1alpha4/handler.go:80]: `v1alpha32` makes me feel confused.

File: pkg/kapis/devops/v1alpha4/handler.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	return func(object runtime.Object) interface{} {
+		pr := object.(*v1alpha4.PipelineRun)
+		runStatusJSON := pr.Annotations[v1alpha4.JenkinsPipelineRunStatusKey]
+		rawRunStatus := json.RawMessage{}
+		// the error will never happen due to raw message won't be nil, so we ignore the error explicitly
+		_ = rawRunStatus.UnmarshalJSON([]byte(runStatusJSON))
```
Comment (LinuxSuRen) [inline pkg/kapis/devops/v1alpha4/handler.go:86]: what will happen if the `runStatusJSON`was modified manually, and it turns out to an invalid json string.

File: pkg/kapis/devops/v1alpha4/register.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+func addToContainer(o Option, ws *restful.WebService, handler *apiHandler) {
+	ws.Route(ws.GET("/namespaces/{namespace}/pipelines/{pipeline}/pipelineruns").
+		To(handler.listPipelineRuns()).
+		Doc("Get all runs of the specified pipeline").
+		Param(ws.PathParameter("namespace", "Namespace of the pipeline")).
+		Param(ws.PathParameter("pipeline", "Name of the pipeline")).
+		Param(ws.QueryParameter("branch", "The name of SCM reference")).
+		Param(ws.QueryParameter("backward", "Backward compatibility. Default value: true").
```
Comment (LinuxSuRen) [inline pkg/kapis/devops/v1alpha4/register.go:39]: Please explain the compatibility. It's hard to understand if users try to read the API document which was generated by this.

**Review by LinuxSuRen — COMMENTED** [review]
I left a few suggestions. Please take a look it.

File: pkg/kapis/devops/v1alpha4/handler.go (hunk 3 above)
Comment (JohnNiang) [reply pkg/kapis/devops/v1alpha4/handler.go:86]: How about returning `{}` if the JSON is invalid?

Comment (LinuxSuRen) [reply pkg/kapis/devops/v1alpha4/handler.go:86]: No matter what, cannot eat this error.

Comment (JohnNiang) [reply pkg/kapis/devops/v1alpha4/handler.go:86]: Please see the first element below:

```json
{
 "items": [
  {},
  {
   "causes": [
    {
     "_class": "io.jenkins.blueocean.service.embedded.rest.AbstractRunImpl$BlueCauseImpl",
     "shortDescription": "Started by user admin",
     "userId": "admin",
     "userName": "admin"
    }
   ],
   "durationInMillis": 62234,
   "enQueueTime": "2021-09-08T21:57:55.633+08:00",
   "endTime": "2021-09-08T21:58:57.876+08:00",
   "estimatedDurationInMillis": 57378,
   "id": "5",
   "organization": "jenkins",
   "pipeline": "main",
   "replayable": true,
   "result": "SUCCESS",
   "runSummary": "stable",
   "startTime": "2021-09-08T21:57:55.642+08:00",
   "state": "FINISHED",
   "type": "WorkflowRun"
  },
  {
   "causes": [
    {
     "_class": "io.jenkins.blueocean.service.embedded.rest.AbstractRunImpl$BlueCauseImpl",
     "shortDescription": "Started by user admin",
     "userId": "admin",
     "userName": "admin"
    }
   ],
   "durationInMillis": 64198,
   "enQueueTime": "2021-09-08T14:22:14.762+08:00",
   "endTime": "2021-09-08T14:23:18.969+08:00",
   "estimatedDurationInMillis": 62034,
   "id": "4",
   "organization": "jenkins",
   "pipeline": "main",
   "replayable": true,
   "result": "SUCCESS",
   "runSummary": "stable",
   "startTime": "2021-09-08T14:22:14.771+08:00",
   "state": "FINISHED",
   "type": "WorkflowRun"
  }
 ],
 "totalItems": 3
}
```

Comment (LinuxSuRen) [reply pkg/kapis/devops/v1alpha4/handler.go:86]: It's not reasonable. There are 3 items. But one of them is empty.

Comment (LinuxSuRen) [reply pkg/kapis/devops/v1alpha4/handler.go:86]: Not necessary to return the message of invalid data to API response. Please use a logger to print it.

Comment (JohnNiang) [reply pkg/kapis/devops/v1alpha4/handler.go:86]: Consider that there is no Jenkins run status yet, what data should be returned

Comment (LinuxSuRen) [reply pkg/kapis/devops/v1alpha4/handler.go:86]: I prefer to return nothing. In this case, the total number should be 2.

**Review by LinuxSuRen — COMMENTED** [review]
Consider there's no strong dependency relation between this PR and https://github.com/kubesphere/kubesphere/pull/4232

I believe that we can merge them without a specific order.

/lgtm

**Pull request merged**